meta-mbl-distro/scripts/lib/mbl/update/payloadimage.py

```diff
@@ -10,12 +10,47 @@
 import shutil
 import tarfile
 
+import mbl.update.versionedclassregistry as vcr
 import mbl.util.fileutil as futil
 
 
 class PayloadImage(abc.ABC):
     """ABC for creating image files for components in an update payload."""
 
+    @classmethod
+    def register(cls):
```

Comment:
We might be able to make this cleaner by using `subclasshook` so we don't need to call this register method on every subclass?

Reply:
I considered doing something like that, but I don't actually want to register every subclass of `PayloadImage`.

Reply:
@rwalton-arm and I spoke IRL... I'm going look into making the registration automatic again, even if it doesn't use `subclasshook`.

Reply:
I can't think of a good way to do this without making things too complicated.

`subclasshook` doesn't really do what we want - it's for customizing which classes are considered subclasses of an ABC. We could potentially use the list of subclasses of the ABC (as defined by `issubclass`) as a registry, but the logic to retrieve things from the registry would get more complicated, and it would require marking some things that _are_ actually subclasses of the ABC as not subclasses as defined by `issubclass`, which is a bit nasty.

Another way to achieve this would be to create a metaclass that automatically registered the correct classes, but the metaclass of the classes involved is already `abc.ABCMeta`, so we'd have to derive a new metaclass from `abc.ABCMeta` and get that to do registration.

I don't think its worth adding this level of complexity in order to remove the calls to the register function, which I've now renamed to `register_versioned_class` from `register` to avoid confusion with the `register` function provided by `abc.ABCMeta`.

Reply:
I'm wondering if we can just maintain a mapping of payload versions to the class objects that will be used in that payload version, rather than keeping a mapping of image version numbers to payload versions. I think we can then basically get rid of the registration mechanism (which seems a bit convoluted to me) and we still have all mappings in one place for easier maintenance.

Reply:
https://github.com/ARMmbed/meta-mbl/pull/781/commits/955f741f2ed28dec2f89a6ae89c04c815fb06f90